### DISCUSSION TITLE
How to get a frictionless development workflow?

### DISCUSSION THREAD

**Discussion by ghost** [mpi4py/mpi4py#118] [Q&A] [answered]
As a beginner I would like to play around with mpi4py's source code.

How can I do this in an efficient way and what's your development workflow?
How do you handle Cython compilation during development?

I already found how to speed up the C compiler
https://github.com/cython/cython/wiki/FAQ#how-can-i-speed-up-the-c-compilation

**Comment by dalcinl** [comment]
I usually just execute `make`. But `make` will just invoke setup.py, and the distutils/setuptools based build system takes care of rerunning Cython and compiling sources. 

Ideally, you should use `pip install --user -e .`, but for some reason I do not know, it is not working (the `--user` flag is not honored). It should work if you use a virtual environment without the `--user` flag, though.

**Comment by ghost** [comment, marked as answer]
Alright `pip install -e .` worked fine in a venv after an initial `make`.
My `make` run takes nearly a minute to complete.
I have to run `make` and `pip install -e .` every time to apply my changes.
The failing `pip install --user -e .` is a bug in pip https://github.com/pypa/pip/issues/7953

**Reply by dalcinl** [reply]
I believe `make` is not enough. `make` runs a bare `python setup.py build`. You need `python setup.py develop`, that way the *.so extension modules are properly copied from the `build` directory to `src/mpi4py/`.

**Reply by ghost** [reply]
Great! I changed the makefile's build recipe to:
`CFLAGS="-O0 -ggdb" $(PYTHON) setup.py build $(opt) && $(PYTHON) setup.py develop`
It takes about 20 seconds to see my changes.
Mpi4py's code is organized as a single cython module. This means tools like `ccache` don't help and cythonize can't run parallel build jobs.

**Reply by dalcinl** [reply]
Yes, it is a single module, and that's not going to change, everything there has to be under the `mpi4py.MPI` namespace.
